### ISSUE TITLE
Investigate detecting hung print requests (vs. long print jobs)

### ISSUE THREAD

**Issue by jsmeix** [apple/cups#5559]
We (i.e. SUSE) analyzed a customer issue
with a printer device that gets its print jobs via IPP
(the CUPS `ipp` backend submits jobs to the printer)
where sometimes a print job "hangs up".

In such cases it helps to terminate the CUPS `ipp` backend
and when it re-runs it can (re)-submit that same print job
that had "hang up" before.

We analyzed a TCP dump of such a hanging job submission
and got the following communication up to the point where it hangs
(job names, user names and other IDs are replaced by `XXX` or `NNN` here)
```
Frame IPPid command                request/response comment
53829 1     Get-Printer-Attributes request
53836 1     Get-Printer-Attributes response         multiple-document-handling-supported: separate-documents-uncollated-copies
53847 2     Validate-Job           request          job-name: XXX.pdf multiple-document-handling: separate-documents-collated-copies
53849 3     Get-Printer-Attributes requuest         requesting-user-name: XXX requested-attributes: compression-supported
53869 3     Get-Printer-Attributes response         compression-supported: none printer-state: idle
53866 2     Validate-Job           response
53873 4     Get-Jobs               request          requesting-user-name: XXX requested-attributes: job-id
53878 5     Create-Job             request          job-name: XXX.pdf multiple-document-handling: separate-documents-collated-copies
53889 4     Get-Jobs               response         OK but no jobs in response
53894 5     Create-Job             response         job-id: NNN job-state: pending job-state-resons: job-data-insufficient
53942 6     Send-Document          request          job-id: NNN last-document: true compression: gzip
```
The printer never sent back an IPP response to the IPP request in frame 53942.
The printer must send an IPP response, even if we submit wrong/invalid data,
so the root cause is inside the printer.

It seems CUPS waits endlessly for an IPP response
(at least it waits more than one hour).

Therefore my first question:
Is it possible to specify a general timeout
how long CUPS waits for an IPP response?

I think the `ipp` backend URI option `contimeout` does not help here
because I think that timeout is not checked in this case here.

While CUPS is waiting for an IPP response no messages appear
in the CUPS log file (at least none with `LogLevel debug`).
This is how the CUPS log file looks up to the point where it hangs
```
D ... Sending file using HTTP/1.1 chunking...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 12100 bytes...
```
For comparison how the CUPS log file looks when things work
after the "hanging" CUPS `ipp` backend was terminated
and when it re-runs it can (re)-submit that same print job
that had "hang up" before
```
D ... Sending file using HTTP/1.1 chunking...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 16384 bytes...
D ... Read 12100 bytes...
D ... Print-Job: successful-ok-ignored-or-substituted-attributes ...
D ... Print job accepted - job ID NNN.
```

My second question:
Is it possible to log messages from time to time
while CUPS is waiting for an IPP response?

My third question is about user notification in such cases:
Is it possible to somehow show to the user 
that a print job cannot proceed
because CUPS is waiting for an IPP response?

**Comment by michaelrsweet** [comment]
Will investigate but in general having arbitrary send timeouts is bad - if the printer runs out of paper then you don’t want to give up, for example.

**Comment by jsmeix** [comment]
@michaelrsweet 

thank you for having a look if things could be improved.

I do fully agree that a default timeout must be very big
like the one week default `contimeout = 7 * 24 * 60 * 60` (in seconds).

From the past I remember an issue we (at SUSE) had with LPRng
where a customer submitted intentionally on Friday evening
huge print jobs knowingly that over the weekend his printer
will run out of paper and he expected that on Monday he could
add paper and his printer will continue to print out the rest.
But a too short default timeout in LPRng had canceled his job
(I think the default timeout was 8 hours or something like this).
The solution was to specify a sufficiently big value for that timeout
(in LPRng the user could specify almost all and everything as needed).

To make it more clear what I mean cf. what I had written at
https://bugzilla.opensuse.org/show_bug.cgi?id=337794#c16
(excerpts as needed here):
```
As long as the backend cannot establish a communication with its
recipient, it cannot cause damage when it loops infinitely and
retries again and again (with a reasonable sleep time between
each retry)
...
with reasonable stderr-messages so that user is continuously
informed for each re-try what goes on (e.g. "123th re-try to
connect to USB printer ACME FunPrinter 1000: no response").
If the user doesn't like it, he can simply cancel his print job.
...
But if there is a communication error while the backend sends data,
it should exit with exit code 3 (CUPS_BACKEND_HOLD) so that the
partially printed job is neither lost nor automatically re-printed.
The user can then decide what to do with his partially printed job.
```

As far as I see the first kind of issues while the backend is about
to establish a communication with its recipient is currently
handled in the CUPS backends via things like `contimeout`.

But I think the second kind of issues when the communication
with the recipient is already successfully established
is currently not yet handled well in the CUPS.

This issue here is about the second kind of issues.

There are several timout settings in cupsd.conf in particular
```
Timeout seconds
  Specifies the HTTP request timeout.
  The default is "300" (5 minutes).
```
where I find in
https://github.com/apple/cups/blob/master/scheduler/cupsd.h#L83
```
#define DEFAULT_TIMEOUT	300 /* Timeout during requests/updates */
```
which sems to indicate that the HTTP request timeout value
should also apply "inside" an established IPP communication
because IPP runs "on top of" HTTP.

Accordingly it seems by default after 5 minutes without any response
to the Send-Document request "something" should have happened
but actually at least nothing visible had happened.

I don't know about the IPP/HTTP implementation details in CUPS.
Perhaps there is alreary a silent loop (looping each 5 minutes by default)
that waits endlessly for a response to the Send-Document request?

If yes, all what would be missing is to make that waiting loop
more verbose so that the user is continuously informed
for each re-try what is going on e.g.
```
123th time waiting 300 seconds for Send-Document request to complete: no response
```
The user can then decide what to do with his stuck print job.

**Comment by michaelrsweet** [comment]
@jsmeix The 300 second (5 minute) timeout in cupsd.conf limits how long cupsd waits for a client to send more request data. There is a similar keep-alive timeout for idle connections, smaller timeouts for reading of TLS and IPP values, etc. None of those apply here.

The information you have provided indicates that the printer is hanging during the Send-Document request, but using Print-Job instead works which points to a printer bug. Like I said, we need to be EXTREMELY conservative/careful timing out these sorts of requests because it is easy to get a false positive and resend a large job...

I'll keep this bug open to track potential solutions, but don't expect a quick one... :/

**Comment by jsmeix** [comment]
Sometimes the printer hangs up during the Send-Document request
(the printer never sends an IPP response to the Send-Document request)
but some time later the exact same procedure works
(by resubmitting the exact same print job via restart of the `ipp` backend).
This is clearly a printer bug, cf. my initial
https://github.com/apple/cups/issues/5559#issue-430382976
```
The printer never sent back an IPP response to the IPP request ...
The printer must send an IPP response, even if we submit wrong/invalid data,
so the root cause is inside the printer.
```

I do fully agree that CUPS needs to be extremely conservative/careful
timing out these sorts of requests because it is easy to get a false positive
and resend a large job, cf. my
https://github.com/apple/cups/issues/5559#issuecomment-481617184
where I wrote about an issue that I remember from the past
```
where a customer submitted intentionally on Friday evening
huge print jobs knowingly that over the weekend his printer
will run out of paper and he expected that on Monday he could
add paper and his printer will continue to print out the rest
```
So any kind of default timeout must be very big (several days)
so that such a timeout is in practice useless (i.e. in practice
the default behaves as if the timeout was unlimited).

Therefore - from my point of view - the actual issue in CUPS
is not to timeout requests but that CUPS is silent while it is waiting
an extradordinary long time during an ongoing IPP request,
cf. what I wrote above (excerpts):
```
While CUPS is waiting for an IPP response no messages appear
in the CUPS log file (at least none with LogLevel debug).
...
My second question:
Is it possible to log messages from time to time
while CUPS is waiting for an IPP response?

My third question is about user notification in such cases:
Is it possible to somehow show to the user
that a print job cannot proceed
because CUPS is waiting for an IPP response?
```
```
... all what would be missing is to make that waiting loop
more verbose so that the user is continuously informed
for each re-try what is going on e.g.

123th time waiting 300 seconds for Send-Document request to complete: no response

The user can then decide what to do with his stuck print job.
```

In other words, simply put:
CUPS should keep the user informed what is going on with his print job
so that user can decide what to do if his print job is stuck.

**Comment by jsmeix** [comment]
The changes in CUPS in https://github.com/apple/cups/issues/5570
```
Increase the MultipleOperationTimeOut default to 15 minutes
...
Bump the default Timeout to 15 minutes as well
```
are good examples that in practice the default timeout values
must be rather longer than shorter to avoid unintended interruption
of valid ongoing tasks that take a long time, cf.
https://github.com/apple/cups/issues/5570#issuecomment-488374265
```
... a single page (ergo single PDF) that had 25 little images on it ...
That's why it took so long to produce.
```

**Comment by michaelrsweet** [comment]
Tracking in OpenPrinting CUPS.
